integrations/kube-agent-updater/pkg/controller/statefulset.go

```diff
@@ -0,0 +1,284 @@
+package controller
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type StatefulSetVersionUpdater struct {
+	VersionUpdater
+	kclient.Client
+	Scheme *runtime.Scheme
+}
+
+// Reconcile treats a reconciliation request for a StatefulSet object. It gets the
+// object, retrieves its current version, and calls the VersionUpdater to find
+// if the Deployment should be updated. If it's the case, it changes the
+// Teleport image version and updates the Deployment in Kubernetes.
+//
+// WARNING: StatefulSets rollouts can end up being stuck because of unready pods.
```

Comment:
What happens if reconciliation takes an eternity? Does the Kube API wait for the operation to happen? 
Can't we spin a goroutine that waits for the rollout to happen and revert it if the new pods are unhealthy?

Reply:
The controller is not waiting for the rollout. The deletion could hang tough. I'll have to test to see if we can stall the controller by having pods that can't be deleted. If this happens, I suppose we'll need to add a timeout on the deletion call if that isn't the case already.

If we rollout a new version and it breaks, we cannot know if new pods are not ready because of the new version or something else (like expired state) and might block updates for an issue which is not update-related. If we were to rollback, we would need to keep track of which version has been ruled out to avoid flickering back and forth. If we get this mechanism wrong we'll end up with agents stuck with an old version, which we want to avoid at all costs. There are definitely cases in which this design will cause disruption (e.g.: the new agent version does not work, we update the sts, the rollout is stuck, a nodepool update happens, all healthy pods are removed). I understood the RFD spirit as "disruption is acceptable as long as we can rollout a fix", I don't think there's a safe way to do sts rollback while ensuring no lock can happen.

Reply:
When editing the statefulset spec and we expect the pods to be updated (first it's the N-1, n-2...0) and something wrong happens, we might have a very weird behavior of Kube controller restarting all the pods (even the pods that didn't upgrade before the rollback) and everything in the cluster can break just because Kube controller assumes it should operate 2 rolling update requests.

Using the rolling update strategy with rollback informs the controller that the operation must stop immediately and the statefulset spec must be reverted to a prior state. It means that if replica 3 fails to pull the image/config change, replica 0-2 won't be restarted in case of a rollback.

The updater can always check for active rolling updates when it starts but I feel it would be the ideal scenario here since it avoids scenarios where connection to auth fails and the cluster becomes inaccessible

Reply:
> When editing the statefulset spec and we expect the pods to be updated (first it's the N-1, n-2...0) and something wrong happens, we might have a very weird behavior of Kube controller restarting all the pods (even the pods that didn't upgrade before the rollback) and everything in the cluster can break just because Kube controller assumes it should operate 2 rolling update requests.

I don't understand how this is possible; I'll take some time to understand how the statefulset controller works. Do you mind if I call you later so you can explain to me how you encountered such behaviour?

Reply:
We discussed the associated risk and as the statefulset controller implements rollback as a rollforward, we can roll forward without any rollback logic.

As the auth server might be down (because of a maintenance) for 10 minutes, we'll want to protect the other pods from being killed (e.g. during a node rollout/downscale/drain) while the new pod is not ready. This implies running multiple pods and enabling the pod disruption budget, this will be done in [Teleport Discover](https://github.com/gravitational/teleport/issues/22731).